src/ring_theory/adjoin_root.lean

```diff
@@ -254,6 +266,31 @@ end power_basis
 
 section equiv
 
+section integral_domain
+
+variables [integral_domain R] [integral_domain S] [algebra R S]
+variables (g : polynomial R) (pb : _root_.power_basis R S)
+
+/-- If `S` is an extension of `R` with power basis `pb` and `g` is a monic polynomial over `R`
+such that `pb.gen` has a minimal polynomial `g`, then `S` is isomorphic to `adjoin_root g`.
+
+Compare `power_basis.equiv'`, which would require `h₂ : aeval pb.gen (minpoly R (root g)) = 0`;
+that minimal polynomial is not guaranteed to be identical to `g`. -/
+def equiv' (h₁ : aeval (root g) (minpoly R pb.gen) = 0) (h₂ : aeval pb.gen g = 0) :
+  adjoin_root g ≃ₐ[R] S :=
+{ to_fun := adjoin_root.lift_hom g pb.gen h₂,
+  inv_fun := pb.lift (root g) h₁,
+  left_inv := λ x, induction_on g x $ λ f, by rw [lift_hom_mk, pb.lift_aeval, aeval_eq],
+  right_inv := λ x, begin
+    obtain ⟨f, hf, rfl⟩ := pb.exists_eq_aeval x,
+    rw [pb.lift_aeval, aeval_eq, lift_hom_mk]
+  end,
+  .. adjoin_root.lift_hom g pb.gen h₂ }
+
```

Comment:
Should we also add these:
```suggestion
@[simp] lemma coe_equiv' (h₁ : aeval (root g) (minpoly R pb.gen) = 0) (h₂ : aeval pb.gen g = 0) :
  (equiv' g pb h₁ h₂ : adjoin_root g → S) = adjoin_root.lift_hom g pb.gen h₂ :=
rfl

@[simp] lemma coe_equiv'_symm (h₁ : aeval (root g) (minpoly R pb.gen) = 0) (h₂ : aeval pb.gen g = 0) :
  ((equiv' g pb h₁ h₂).symm : S → adjoin_root g) = pb.lift (root g) h₁ :=
rfl
```
and maybe also some (non-simp) `_apply` and `_symm_apply` lemmas?

Reply:
I don't think the two `coe_equiv'` lemmas are quite what we want, because the two directions are also `alg_hom`s. So I went with `@[simps]`ing the definition and adding `to_alg_hom` lemmas.

Reply:
I'm probably being dense, but can you elaborate on why are the lemmas I stated less desirable?

The job I intended these lemmas to do was to enable `simp` to determine the definition `equiv' g pb h₁ h₂` when it is coerced to a function (and its `symm`).

Wouldn't `_apply` lemmas generated by `simps` be consequences of these? As in:
```lean
@[simp] lemma coe_equiv' (h₁ : aeval (root g) (minpoly R pb.gen) = 0) (h₂ : aeval pb.gen g = 0) :
  (equiv' g pb h₁ h₂ : adjoin_root g → S) = adjoin_root.lift_hom g pb.gen h₂ :=
rfl

-- This is the `equiv'_apply` lemma that `simps` creates:
example (h₁ : aeval (root g) (minpoly R pb.gen) = 0) (h₂ : aeval pb.gen g = 0) (x : adjoin_root g) :
  equiv' g pb h₁ h₂ x = adjoin_root.lift_hom g pb.gen h₂ x :=
by simp
```